### PR TITLE
CASM-3717 implement hooks

### PR DESCRIPTION
**Pull request by atifsyedali**
## Summary and Scope

Implements hooks in IUF.

1. Refactored the sessions.go code into smaller files and more containable corresponding *_test.go files.
2. There is a corresponding docs-csm change that creates the Argo template for the master_host execution context to run the hook script as part of the Argo workflow.
3. Added a lot of unit tests to cover the new functionality and some of existing functionality in case I didn't break anything during refactor.

_Is this change backwards incompatible, backwards compatible, or a backwards compatible bugfix?_

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [CASM-3717](https://jira-pro.its.hpecorp.net:8443/browse/CASM-3717)
* Change will also be needed in https://github.com/Cray-HPE/docs-csm/pull/2954

## Testing

Will be testing on frigg today.

### Tested on:

  * `<development system>`
  * Local development environment

### Test description:

Lots of unit tests

## Risks and Mitigations

No

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [ ] CHANGELOG.md updated
- [ ] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

**Comment by atifsyedali** [comment]
@haasken-hpe FYI, I am going to be writing other PRs rebased on top of this one so I can do simultaneous testing on frigg.

File: .version [hunk 1]
```diff
@@ -1 +1 @@
-0.9.6
+0.9.7
```
Comment (haasken-hpe) [inline .version:1]: If this particular PR is going to be merged, then this version needs to be fixed as `0.9.7` has already been used. If #137 will be used instead, and this PR is abandoned, then it's already handled in that PR.

File: charts/v1.0/cray-nls/Chart.yaml [hunk 2]
```diff
@@ -24,7 +24,7 @@
 
 apiVersion: v2
 name: "cray-nls"
-version: 1.4.38
+version: 1.4.39
```
Comment (haasken-hpe) [inline charts/v1.0/cray-nls/Chart.yaml:27]: Same comment as above.

**Review by haasken-hpe — APPROVED** [review]

**Comment by atifsyedali** [comment]
Consumed and delivered as part of https://github.com/Cray-HPE/cray-nls/pull/137

**Pull request closed without merging**